### PR TITLE
MongoDB plugin implementation

### PR DESCRIPTION
**Pull request by vpavliv**
MongoDB plugin to store blocks.

### CODE REVIEW

File: libraries/plugins/mongo_db/mongo_db_writer.cpp [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+        try {
+            uri = mongocxx::uri {uri_str};
+            mongo_conn = mongocxx::client {uri};
+            db_name = uri.database().empty() ? "Golos" : uri.database();
+            ilog("MongoDB plugin initialized.");
+        }
+        catch (mongocxx::exception & ex) {
+            ilog(ex.what());
+            throw;
```
Comment (afalaleev) [inline libraries/plugins/mongo_db/mongo_db_writer.cpp]: Don't throw an unknown for golosd exception.
If you need throw an exception - declare new type based on fc::exception and throw it.

Comment (vpavliv) [reply libraries/plugins/mongo_db/mongo_db_writer.cpp]: You right. In fact, i guess if mongo plugin would be broken its not the end of the world, we can just log the exception and get on with our lives.

**Comment by vpavliv** [comment]
Operations are not complete yet just half of them.

**Comment by vpavliv** [comment]
Alexander you were asking for a sample from my MongoDB, here you are

{ "_id" : ObjectId("5a987fa25ae2151e640c10cb"), "block_num" : "58", "block_id" : "0000003aeb0255247d178ca13af053f3272da8b1", "prev_block_id" : "000000392c380947b297c1ae0ee5bf2ce79f5cee", "timestamp" : "2016-10-18T11:04:39", "witness" : "cyberfounder", "created_at" : "2018-03-01T22:33:06", "Transactions" : [ { "id" : "55694b205ec141963b3353ec3439ee429b6c1d27", "ref_block_num" : "57", "expiration" : "2018-03-01T22:33:06", "Operations" : [ { "pow2" : { "props" : { "account_creation_fee" : "0.001 GOLOS", "maximum_block_size" : "131072", "sbd_interest_rate" : "1000" }, "new_owner_key" : "GLS6f4SV56MbSFqBSqo3tbf9ucH6gepsHECDnFPQvcMgUw2JuC8Cw" } } ] } ] }
{ "_id" : ObjectId("5a987fa35ae2151e640c1159"), "block_num" : "200", "block_id" : "000000c8b706dc5e38b3a9f181129a9f9403e55b", "prev_block_id" : "000000c77b2475392690a2c503b268706066ed17", "timestamp" : "2016-10-18T11:12:45", "witness" : "gfunk", "created_at" : "2018-03-01T22:33:07", "Transactions" : [ { "id" : "d233796e0c6512fbef1fd00b9d829f98746b28e5", "ref_block_num" : "199", "expiration" : "2018-03-01T22:33:07", "Operations" : [ { "account_witness_vote" : { "account" : "testz", "witness" : "testzcrypto-04", "approve" : true } } ] } ] }

File: plugins/mongo_db/mongo_db_operations.cpp [hunk 2]
```diff
@@ -0,0 +1,819 @@
...
+        for (auto iter : signature) {
+            retVal += iter;
+        }
+        return retVal;
+    }
+    /////////////////////////////////////////////////
+
+
+    document write_operation(const operation& op) {
+        return document{};
```
Comment (vpavliv) [inline plugins/mongo_db/mongo_db_operations.cpp]: Guys i have a question. How do i extract proper operation from static_variant? So far the only thought i got is to cal .where() - it is a position of type that is currently in operation, and then to deduce type from declaration of operation static variant., and then cal .as<Type>() and get what i want. This does not feels like right tho. Any ideas?

**Comment by gropox** [comment]
Если можно, то добавьте возможность раздельного харанения суммы и наименование ассета в таких операциях как transfer, author_rewards и других. Это позволит агрегировать данные средствами mongodb.

File: plugins/mongo_db/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  # The *.cmake package files provided by mongo-cxx-driver don't give us the
+  # absolute path to the libraries, which is needed whenever they are not
+  # installed in system-known locations.  CMake requires the absolute paths
+  # in target_link_libraries() since we are builiding an archive and the
+  # link step for all executables using this archive must include the
+  # mongo-cxx-driver libraries libmongocxx and libbsoncxx.
+
+  find_package(libbsoncxx REQUIRED)
+  message(STATUS "Found bsoncxx headers: ${LIBBSONCXX_INCLUDE_DIRS}")
+  message(STATUS "Searching for bsoncxx library ${LIBBSONCXX_LIBRARIES} in ${LIBBSONCXX_LIBRARY_DIRS}")
```
Comment (afalaleev) [inline plugins/mongo_db/CMakeLists.txt]: Replace message to: Found bsoncxx library: ${LIBBSONCXX_LIBRARIES}

File: plugins/mongo_db/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  find_library(GOLOS_LIBBSONCXX ${LIBBSONCXX_LIBRARIES}
+               PATHS ${LIBBSONCXX_LIBRARY_DIRS} NO_DEFAULT_PATH)
+  message(STATUS "Found bsoncxx library: ${GOLOS_LIBBSONCXX}")
```
Comment (afalaleev) [inline plugins/mongo_db/CMakeLists.txt]: Remove these 3 lines: from find_library() to message()

File: plugins/mongo_db/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+  find_package(libmongocxx REQUIRED)
+  message(STATUS "Found mongocxx headers: ${LIBMONGOCXX_INCLUDE_DIRS}")
+  message(STATUS "Searching for mongocxx librariy ${LIBMONGOCXX_LIBRARIES} in ${LIBMONGOCXX_LIBRARY_DIRS}")
```
Comment (afalaleev) [inline plugins/mongo_db/CMakeLists.txt]: Replace message to: Found mongocxx library: ${LIBMONGOCXX_LIBRARIES}

File: plugins/mongo_db/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  find_library(GOLOS_LIBMONGOCXX ${LIBMONGOCXX_LIBRARIES}
+               PATHS ${LIBMONGOCXX_LIBRARY_DIRS} NO_DEFAULT_PATH)
+  message(STATUS "Found mongocxx library: ${GOLOS_LIBMONGOCXX}")
```
Comment (afalaleev) [inline plugins/mongo_db/CMakeLists.txt]: Remove these 3 lines from find_library() to message()

File: plugins/mongo_db/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+  target_link_libraries(
+          golos_${CURRENT_TARGET}
+          golos::json_rpc
+          golos_chain
+          golos::chain_plugin
+          appbase
+          fc
+          ${GOLOS_LIBBSONCXX}
+          ${GOLOS_LIBMONGOCXX}
```
Comment (afalaleev) [inline plugins/mongo_db/CMakeLists.txt]: Replace ${GOLOS_LIBBSONCXX} with ${LIBBSONCXX_LIBRARIES}.
Replace ${GOLOS_LIBMONGOCXX} with ${LIBMONGOCXX_LIBRARIES}.

**Comment by vpavliv** [comment]
To gropox : Готово. Теперь все ассеты хранятся следующим образом: 
amount: ""
symbol_name: ""

**Comment by vpavliv** [comment]
Data format in mongo collections is the following:
_id
block_num
block_id
block_prev_block_id
block_timestamp
block_witness
block_created_at
Transactions:
transaction_id
transaction_ref_block_num
transaction_expiration
Operations:

**Comment by vpavliv** [comment]
Format of operation collections is the following
Operation specific fields
Block general fields
Transaction general fields

File: plugins/mongo_db/mongo_db_operations.cpp [hunk 8]
```diff
@@ -0,0 +1,1193 @@
...
+        const comment_object* comment_obj_ptr;
+
+        comment_obj_ptr =_db.find_comment(auth, perm);
+
+        if (comment_obj_ptr == NULL) {
+            return;
+        }
+
+        comment_object comment_obj = *comment_obj_ptr;
+
```
Comment (afalaleev) [inline plugins/mongo_db/mongo_db_operations.cpp]: Add `json_metadata` from comment_object

File: plugins/mongo_db/mongo_db_operations.cpp [hunk 9]
```diff
@@ -0,0 +1,1193 @@
...
+                  << "maximum_block_size" << std::to_string(props.maximum_block_size)
+                  << "sbd_interest_rate" << std::to_string(props.sbd_interest_rate);
+        return props_doc;
+    }
+
+    void format_asset(const asset& _asset, document& asset_doc, const std::string& val) {
+        array arr;
+        arr << std::to_string(_asset.to_real());
+        arr << _asset.symbol_name();
+        asset_doc << val << arr;
```
Comment (afalaleev) [inline plugins/mongo_db/mongo_db_operations.cpp]: Split this to:
asset_doc << val + "_value" <<  std::to_string(_asset.to_real());
asset_doc << val + "_symbol << _asset.symbol_name();

File: plugins/mongo_db/mongo_db_operations.cpp [hunk 10]
```diff
@@ -0,0 +1,1249 @@
...
+        return props_doc;
+    }
+
+    void format_asset(const asset& _asset, document& asset_doc, const std::string& val) {
+
+        asset_doc << val + "_value" << std::to_string(_asset.to_real());
+        asset_doc << val + "_symbol" << _asset.symbol_name();
+    }
+
+    void format_value(document& doc, const std::string& name, const std::string& value) {
```
Comment (afalaleev) [inline plugins/mongo_db/mongo_db_operations.cpp]: Should allow to store integers/doubles/...  Not only strings. It is impossible to aggregate data in database if all fields is strings.

File: plugins/mongo_db/mongo_db_operations.cpp [hunk 11]
```diff
@@ -0,0 +1,1249 @@
...
+        if (name.empty() || value.empty()) {
+            return;
+        }
+        doc << name << value;
+    }
+
+    std::string to_string(const signature_type& signature) {
+        std::string retVal;
+
+        for (auto iter : signature) {
```
Comment (afalaleev) [inline plugins/mongo_db/mongo_db_operations.cpp]: `auto iter` -> `auto& iter`

File: plugins/mongo_db/mongo_db_operations.cpp [hunk 12]
```diff
@@ -0,0 +1,1249 @@
...
+            format_asset(comment_obj.total_payout_value, comment_doc, "total_payout_value");
+            // total_pending_payout_value
+            format_value(comment_doc, "total_vote_weight", std::to_string(comment_obj.total_vote_weight));
+            // url
+            format_value(comment_doc, "vote_rshares", std::to_string(comment_obj.vote_rshares.value));
+            // last_reply
+            // last_reply_by
+            format_value(comment_doc, "json_metadata", comment_obj.json_metadata.c_str());
+
+            format_comment_active_votes(comment_obj, comment_doc);
```
Comment (afalaleev) [inline plugins/mongo_db/mongo_db_operations.cpp]: Should be stored as independent collection.

File: plugins/mongo_db/mongo_db_operations.cpp [hunk 13]
```diff
@@ -0,0 +1,1249 @@
...
+                case second_payout:
+                    comment_mode = "second_payout";
+                    break;
+                case archived:
+                    comment_mode = "archived";
+                    break;
+            }
+
+            format_value(comment_doc, "mode", comment_mode);
+            format_value(comment_doc, "net_rshares", std::to_string(comment_obj.net_rshares.value));
```
Comment (afalaleev) [inline plugins/mongo_db/mongo_db_operations.cpp]: int64_t !

File: plugins/mongo_db/mongo_db_operations.cpp [hunk 14]
```diff
@@ -0,0 +1,1249 @@
...
+            format_value(comment_doc, "net_votes", std::to_string(comment_obj.net_votes));
```
Comment (afalaleev) [inline plugins/mongo_db/mongo_db_operations.cpp]: int64_t !

File: plugins/mongo_db/mongo_db_operations.cpp [hunk 15]
```diff
@@ -0,0 +1,1249 @@
...
+            format_value(comment_doc, "total_vote_weight", std::to_string(comment_obj.total_vote_weight));
+            // url
+            format_value(comment_doc, "vote_rshares", std::to_string(comment_obj.vote_rshares.value));
+            // last_reply
+            // last_reply_by
+            format_value(comment_doc, "json_metadata", comment_obj.json_metadata.c_str());
+
+            format_comment_active_votes(comment_obj, comment_doc);
+
+            // TODO: the following fields depends on follow-plugin operations - custom_json ...
```
Comment (afalaleev) [inline plugins/mongo_db/mongo_db_operations.cpp]: Export as independent collection.

File: plugins/mongo_db/mongo_db_operations.cpp [hunk 16]
```diff
@@ -0,0 +1,1249 @@
...
+            format_value(comment_doc, "body_length", std::to_string(comment_obj.body.length()));
+            format_value(comment_doc, "cashout_time", comment_obj.cashout_time.to_iso_string());
+            format_value(comment_doc, "category", comment_obj.category.c_str());
+            format_value(comment_doc, "children", std::to_string(comment_obj.children));
+            format_value(comment_doc, "children_abs_rshares", std::to_string(comment_obj.children_abs_rshares.value));
+            format_value(comment_doc, "children_rshares2", comment_obj.children_rshares2);
+            format_value(comment_doc, "created", comment_obj.created.to_iso_string());
+            format_asset(comment_obj.curator_payout_value, comment_doc, "curator_payout_value");
+            format_value(comment_doc, "depth", std::to_string(comment_obj.depth));
+            format_value(comment_doc, "_id", std::to_string(comment_obj.id._id));
```
Comment (afalaleev) [inline plugins/mongo_db/mongo_db_operations.cpp]: Should be stored as sha(author, perm)

**Comment by afalaleev** [comment]
Cancel this pull request. 
A new pull request will be without so long history.

**Pull request closed without merging**